### PR TITLE
Feature/migrate tweeks

### PR DESCRIPTION
**Pull request by stefanhar**
- Added locale to migrate script
- Broke script up into multiple files
- Made script work without AWS connection
- Added logs
- "Fixed" circular JSON error
- Added readme
- Added migrate to workspaces

### CODE REVIEW

**Review by haukurh — CHANGES_REQUESTED** [review]
looks really good, works very well on local docker ES (although I didn't test AWS functionality)

only minor changes requested:
fix aws package name + minor typos

File: apps/services/search-indexer/src/migrate/aws.ts [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+  })
+}
+
+// construct the name used for packages in AWS ES
+const createPackageName = (
+  locale: string,
+  analyzerType: string,
+  version: string,
+) => {
+  return `${locale}-${analyzerType}-${version}-test1`
```
[the remark this answers is not in the document]
Comment (stefanhar) [reply apps/services/search-indexer/src/migrate/aws.ts:24]: Damn

File: apps/services/search-indexer/src/migrate/aws.ts [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+    })
+    .promise()
+
+  const foundPackage = domainPackageList.DomainPackageDetailsList.find(
+    (listItem) => listItem.PackageID === packageId,
+  )
+
+  return {
+    packageStatus: packages.PackageDetailsList[0].PackageStatus,
+    domainStatus: foundPackage?.DomainPackageStatus ?? 'DISSOCIATED',
```
[the remark this answers is not in the document]
Comment (stefanhar) [reply apps/services/search-indexer/src/migrate/aws.ts:67]: Cause it disapears from the domainPackage list.
I will update the name here to make that more clear

File: apps/services/search-indexer/src/migrate/aws.ts [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+        S3Key: createS3Key(analyzerType, locale),
+      },
+    }
+
+    logger.info('Creating AWS ES package', params)
+
+    const esPackage = await awsEs.createPackage(params).promise()
+
+    // we have to wait for package to be ready cause AWS ES can only process one request at a time
+    await waitForPackageStatus(esPackage.PackageDetails.PackageID, 'AVAILABLE')
```
[the remark this answers is not in the document]
Comment (stefanhar) [reply apps/services/search-indexer/src/migrate/aws.ts:282]: Nice catch will add the correct types to `waitForPackageStatus `

File: apps/services/search-indexer/src/migrate/aws.ts [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+
+  // wait for all request to resolve
+  await Promise.all(responses)
+  return true
+}
+
+/*
+createAwsEsPackagesshould only run when we are updating, we can therefore assume no assigned packages exist in AWS ES for this version
+We run a remove packages for this verson function to handle failed partial updates
+*/
```
[the remark this answers is not in the document]
Comment (stefanhar) [reply apps/services/search-indexer/src/migrate/aws.ts:251]: Can only find "verson"

[the remark this answers is not in the document]
Comment (haukurh) [reply apps/services/search-indexer/src/migrate/aws.ts:251]: not very helpful comment from me, but also missing space for "createAwsEsPackagesshould"

File: apps/services/search-indexer/src/migrate/elastic.ts [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+
+  logger.info('Created template config', { locale })
+  return config
+}
+
+export const getEsTemplate = async (locale: string) => {
+  const templateName = getTemplateName(locale)
+  logger.info('Geting old template', { templateName })
+  const client = await esService.getClient()
+  const templateBody = await client.indices
```
[the remark this answers is not in the document]
Comment (stefanhar) [reply apps/services/search-indexer/src/migrate/elastic.ts:113]: Fixed

File: apps/services/search-indexer/src/migrate/elastic.ts [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+) => {
+  logger.info('Moving alias to new version', {
+    newIndexVersion,
+    oldIndexVersion,
+  })
+  const aliasName = getIndexBaseName(locale)
+  const oldIndexName = getIndexNameForVersion(locale, oldIndexVersion)
+  const newIndexName = getIndexNameForVersion(locale, newIndexVersion)
+
+  // this is the recomented way to rename an alias, this operation is atomic when executed in a single api request
```
[the remark this answers is not in the document]
Comment (stefanhar) [reply apps/services/search-indexer/src/migrate/elastic.ts:251]: Fixed

File: libs/api/content-search/src/services/elastic.service.ts [hunk 7]
```diff
@@ -207,9 +207,9 @@ export class ElasticService {
 
   async ping() {
     const client = await this.getClient()
-    return client.ping().catch((e) => {
-      ElasticService.handleError('Error in ping', {}, e)
-    })
+    const result = await client.ping()
+    logger.info('Got elasticsearch ping response', { result })
```
[the remark this answers is not in the document]
Comment (stefanhar) [reply libs/api/content-search/src/services/elastic.service.ts:211]: It was like that b4 refactor in the migration, will remove

File: apps/services/search-indexer/src/app/indexing.service.ts [hunk 8]
```diff
@@ -152,12 +152,17 @@ export class IndexingService {
     // related articles has a recursive nesting problem, we prune it for now
     if (entry.fields?.relatedArticles?.[0]?.fields) {
       logger.info('Removing nested related articles from related articles')
```
[the remark this answers is not in the document]
Comment (stefanhar) [reply apps/services/search-indexer/src/app/indexing.service.ts:154]: will remove

File: apps/services/search-indexer/src/migrate/aws.ts [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+  totalSecondsWaited = 0,
+) => {
+  const secondsBetweenRequests = 5
+  const timeoutSeconds = 300
+
+  if (totalSecondsWaited >= timeoutSeconds) {
+    throw new Error(`Failed to get status for package ${packageId}`)
+  }
+
+  // if we find the desired status in eather the domain package list or the unassigned package list we assume success
```
[the remark this answers is not in the document]
Comment (stefanhar) [reply apps/services/search-indexer/src/migrate/aws.ts]: Fixed

File: apps/services/search-indexer/src/migrate/aws.ts [hunk 10]
```diff
@@ -0,0 +1,396 @@
+import { logger } from '@island.is/logging'
+import AWS from 'aws-sdk'
+import { PutObjectRequest } from 'aws-sdk/clients/s3'
+import { environment } from '../environments/environment'
+import _ from 'lodash'
```
[the remark this answers is not in the document]
Comment (stefanhar) [reply apps/services/search-indexer/src/migrate/aws.ts]: Removed

File: apps/services/search-indexer/src/migrate/aws.ts [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+  await sleep(secondsBetweenRequests)
+
+  return await waitForPackageStatus(
+    packageId,
+    desiredStatus,
+    (totalSecondsWaited = totalSecondsWaited + secondsBetweenRequests),
+  )
+}
+
+// checks connection and valdiates that we have access to requested domain
```
[the remark this answers is not in the document]
Comment (stefanhar) [reply apps/services/search-indexer/src/migrate/aws.ts]: Fixed

File: apps/services/search-indexer/src/migrate/aws.ts [hunk 12]
```diff
@@ -0,0 +1,396 @@
...
+  )
+
+  // we remove packages that are in use from allAwsEsPackageIds and return
+  return allAwsEsPackageIds.filter(
+    (awsEsPackageId) => !inUseEsPackageIds.includes(awsEsPackageId),
+  )
+}
+
+export const getUnusedAwsEsPackages = async (
+  inuseEsPackages: AwsEsPackage[],
```
[the remark this answers is not in the document]
Comment (stefanhar) [reply apps/services/search-indexer/src/migrate/aws.ts]: Fixed

File: apps/services/search-indexer/src/migrate/elastic.ts [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+
+// elasticsearch does not provide all return variants in its documentation, we only type a sensible subset of the return object
+interface EsIndex {
+  health: 'green' | 'yellow' | 'red'
+  index: string
+  status: string
+  ['docs.count']: string
+  ['docs.deleted']: string
+}
+const getAllIndice = async () => {
```
[the remark this answers is not in the document]
Comment (stefanhar) [reply apps/services/search-indexer/src/migrate/elastic.ts]: Fixed

File: apps/services/search-indexer/src/migrate/elastic.ts [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+
+export const getCurrentVersionFromIndices = async (
+  locale: string,
+): Promise<number> => {
+  logger.info('Getting index version from elasticsearch indexes', { locale })
+  const indice = await getAllIndice()
+  const indicePrefix = getIndexBaseName(locale)
+
+  // find the highest version matching our index prefix using our custom index naming convention
+  return indice.reduce((higherstVersion, { index }) => {
```
[the remark this answers is not in the document]
Comment (stefanhar) [reply apps/services/search-indexer/src/migrate/elastic.ts]: Fixed

File: libs/api/content-search/src/services/elastic.service.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (haukurh) [reply libs/api/content-search/src/services/elastic.service.ts]: Also one note, why did you remove the ElasticService.handleError ?

[the remark this answers is not in the document]
Comment (stefanhar) [reply libs/api/content-search/src/services/elastic.service.ts]: I don't see a reason, will add it back

**Review by haukurh — APPROVED** [review]
LGTM

**Pull request merged**